treewide: Qt5.X -> Qt5.15 (when possible)

freezeboy opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

I tried to migrate the software still using Qt5.12 to a newer version, sometimes I had to update the software to make it work.

Among the reasons blocking the packages, I already spotted:

 * Dependency to qtwebkit which is broken with Qt5.15
 * Use of QPainterPath class, apparently the header definition has moved between 5.14 and 5.15
 * Some package build configurations forbid deprecated members use ... and there are such in Qt5.15
 * Other oddities in the code that is not more valid (changing, return types)
 * A few of them compile .. then segfault (digitalbitbox)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

freezeboy committed:
digitalbitbox: 2.2.2 -> 3.0.0

freezeboy committed:
qlcplus: 4.12.2 -> 4.12.3

freezeboy committed:
golden-cheetah: use Qt5.14

freezeboy committed:
sdrangel: 4.11.12 -> 4.21.1

freezeboy committed:
plex-media-player: 2.55.0.1069 -> 2.58.0.1076

freezeboy committed:
teamspeak_client: migrate to Qt5.15

freezeboy committed:
protonmail-bridge: migrate to Qt5.15

freezeboy committed:
vnote: 2.8.2 2.8.2 -> 2.10

Also migrate to Qt5.15

freezeboy committed:
tellico: migrate to Qt5.15

freezeboy committed:
calaos_installer: 3.1 -> 3.5

Also migrate to Qt 5.15

freezeboy committed:
qcsxcad: migrate to Qt511 + additional cleanup

 * Split into multiple outputs
 * Rewrite the derivation without stdenv

freezeboy committed:
eagle: migrate to Qt5.15

freezeboy committed:
appcsxcad: migrate to Qt5.15

 * remove useless wrapQtAppsHook

freezeboy committed:
xflr5: migrate to Qt5.15

rewrite without wrapQtAppsHook and stdenv

freezeboy committed:
solarus: move to multiple outputs

motivation: Let solarus-quest-editor require only required bits

freezeboy committed:
samplv1: 0.9.13 -> 0.9.18, migrate to Qt5.15

Rewrite without stdenv

freezeboy committed:
saga: 7.7.0 -> 7.8.1

 * Remove obsolete patches (merged into this release)
 * rewrite without stdenv

freezeboy committed:
pentobi: 18.1 -> 18.3

 * Rewrite without stdenv and wrapQtAppsHook
 * Fetch source from GithHub

freezeboy committed:
chiaki: migrate to Qt5.15

freezeboy committed:
psi: 1.4 -> 1.5

remove useless wrapQtAppsHook and stdenv

freezeboy committed:
psi-plus: migrate to Qt5.15

remove useless wrapQtAppsHook and stdenv

freezeboy committed:
zoom-us: migrate to Qt5.15

freezeboy committed:
webcamoid: migrate to Qt5.15

rewrite without stdenv

freezeboy committed:
virt-manager-qt: migrate to Qt5.15

freezeboy committed:
kdiff3: migrate to Qt5.15

fix homepage

freezeboy committed:
krename: migrate to Qt5.15

freezeboy committed:
kronometer: migrate to Qt5.15

freezeboy committed:
peruse: migrate to Qt5.15

freezeboy committed:
minitube: 3.2 -> 3.6

rewrite without useless wrapQtAppsHook and stdenv

freezeboy committed:
rsibreak: migrate to Qt5.15

freezeboy committed:
birdtray: migrate to Qt5.15

freezeboy committed:
strawberry: migrate to Qt5.15

freezeboy committed:
flameshot: migrate to Qt5.15, fix homepage

freezeboy committed:
ktorrent: migrate to Qt5.15

freezeboy committed:
clipgrab: migrate to Qt5.15

doronbehar reviewed (approved):
This is a tremendous work! Did you test each and every package to build and run right?

Overall I approve, but the commit messages sometimes miss some details of other changes, it might help to get this approved but I wouldn't make this a blocker.

freezeboy commented:
Yes I built and run all the executables, just to check it starts.

For the commit messages, I can see to improve them, at least the ones where an update was needed.

freezeboy committed:
sonic-lineup: migrate to Qt5.15

doronbehar commented:
Well done :clap:.

freezeboy commented:
Most of the effort is done by my cpu and its fan 😏

freezeboy committed:
skrooge: migrate to Qt5.15, enable tests

freezeboy committed:
skanlite: migrate to Qt5.15

freezeboy commented:
I force pushed with more details in the commit messages, please tell me if you need more

doronbehar commented:
Pushed after a rebase to fix merge conflicts. Ideally I think every usage of a pinned `libsForQt5..` should be accompanied with a comment. I counted 129 usages of `libsForQt514.callPackage` that come with no comment explaining them.

freezeboy commented:
Thank you, I will continue to add other apps, and I think we can think to merge this PR, I will start another PR for to continue this process otherwise we will have more and more conflicts

freezeboy committed:
cutecom: fix build with Qt5.15

Used a dangling commit on upstream master branch as it was not released
and the patch was submitted to opensuse instead of upstream

freezeboy committed:
libsForQt5.herqq: fix build with qt5.15

Using a unreleased commit from upstream.

Also moved the derivation to mkLibsForQt5 to let
the application choose its Qt5 version.

Did not put an alias as previous version was broken
since ages.

freezeboy committed:
p4v: comment why it is broken with Qt5.15

freezeboy committed:
ostinato: 0.9 -> 1.1

 * Port to Qt5.15 in the process
 * Cleanup the derivation to use new style

freezeboy committed:
libsForQt5.kde2-decoration: 1.0 -> 1.1

Fixes build for Qt5.15

Split output to mutilple outputs to reduce user closure size

freezeboy committed:
retroshare: move to alias and explain why it is broken

freezeboy committed:
sigil: migrate to Qt5.15 and fix segfault

freezeboy committed:
libsForQt5.qtcurve: fix build with Qt5.15

moved the initial qtcurve package to mkLibsForQt5 function
to decouple from Qt5 version

added an alias qtcurve -> libsForQt5.qtcurve for backward compatibility

add option to disable gtk2 support (still enabled by default)

freezeboy commented:
Ok, I stop for this PR, time for week end :smile:

jonringer reviewed (approved):
diff LGTM
few programs I tried seem to work as intended

cc @ttuegel 
```
https://github.com/NixOS/nixpkgs/pull/102840
1 package marked as broken and skipped:
retroshare

59 packages built:
appcsxcad birdtray calaos_installer chiaki clipgrab cutecom digitalbitbox eagle flameshot golden-cheetah kdiff3 krename kronometer ktorrent libsForQt5.herqq libsForQt5.kde2-decoration libsForQt5.qcsxcad libsForQt5.qtcurve libsForQt512.herqq libsForQt512.kde2-decoration libsForQt512.qcsxcad libsForQt512.qtcurve libsForQt514.herqq libsForQt514.kde2-decoration libsForQt514.qcsxcad libsForQt514.qtcurve minitube openems ostinato pentobi peruse plex-media-player protonmail-bridge psi psi-plus python27Packages.python-csxcad python37Packages.python-csxcad python37Packages.python-openems python38Packages.python-csxcad python38Packages.python-openems qlcplus rsibreak saga samplv1 sdrangel sigil skanlite skrooge solarus solarus-quest-editor sonic-lineup strawberry teamspeak_client tellico virt-manager-qt vnote webcamoid xflr5 zoom-us
```

gnidorah commented on pkgs/data/themes/qtcurve/default.nix:
```diff
@@ -1,32 +1,24 @@
     repo = "qtcurve";
-    rev = version;
-    sha256 = "0sm1fza68mwl9cvid4h2lsyxq5svia86l5v9wqk268lmx16mbzsw";
+    rev = "e534cd04e6acc90c105b492f3fcf99cc6155ba15";
```
As an alternative to using unstable version you could add the following patch to release https://github.com/KDE/qtcurve/commit/44e2a35ebb164dcab0bad1a9158b1219a3ff6504

freezeboy replied:
Well it was the idea, I just have a few other patches that don't look very dangerous.
Do you think it is still important to add just one commit ?

gnidorah replied:
Do as you like, just don't forget about version :smiley:

doronbehar replied:
I too agree with @gnidorah on this - adding a patch is a bit more rigorous and a version change warrants a commit with `<old version> -> <new version>`.

gnidorah commented on pkgs/data/themes/qtcurve/default.nix:
```diff
@@ -1,32 +1,24 @@
+mkDerivation {
   pname = "qtcurve";
-  inherit version;
+  version = "1.9.1";
```
```suggestion
  version = "2020-06-21";
```
For commit https://github.com/KDE/qtcurve/commit/e534cd04e6acc90c105b492f3fcf99cc6155ba15

doronbehar committed:
kmymoney: pin to qt5.14

jonringer commented:
is this still being actively developed?

freezeboy commented:
If you think the state is ok to be merged, then go, I can work on another PR, I just fixed a few things from various reviewers (and thank all of the btw)

doronbehar committed:
mendeley: comment why it needs qt5.14

doronbehar committed:
merkaator: comment why it needs qt5.14

jonringer commented on pkgs/top-level/all-packages.nix:
```diff
@@ -13199,7 +13199,7 @@ in
     # TODO: remove once `udev` is `systemdMinimal` everywhere.
     udev = systemdMinimal;
-  };
+  }
```
```suggestion
  };
```
to fix ofborg

doronbehar commented:
The last 2 commits should be replaced with a rebase.

doronbehar commented:
Github sometimes doesn't notify subscribers when a force-push happens (perhaps only for rebases). I think this is ready now, and worth merging.

jshholland commented:
Should nextcloud-client be included in this? 22167ae45fb27fac1e6aec7d1bf6ce82516fcb12 pinned it to 5.14, but at the moment it segfaults on launch for me. If I switch it to `libsForQt5.callPackage ../applications/networking/nextcloud-client { };` then it launches and runs fine.

jnetod commented:
https://github.com/NixOS/nixpkgs/pull/96309 fixes qtwebkit on 5.15 too.

freezeboy commented:
I propose to merge this pr, there are still many other packages to test/fix with qt5.15 we can open a new pr without problem

doronbehar commented:
I won't mind keep adding packages here, until we get @ttuegel's approval. Perhaps even the qtwebkit could be fixed here, and more qt5.14 pins could be dropped.

doronbehar commented:
I added the qtwebkit patches for qt515.qtwebkit, and they work with e.g `kmymoney`. @freezeboy unless you wish to not broaden the scope of this PR, I won't mind pushing it here.

freezeboy commented:
Feel free to complete this PR, I am working on other packages atm

ttuegel commented on pkgs/applications/editors/sigil/default.nix:
```diff
@@ -24,6 +24,10 @@ mkDerivation rec {
   ];
 
+  prePatch = ''
+    sed -i '6d' src/Resource_Files/bash/sigil-sh_install
```
This is likely to silently break if the package is updated. Let's do something else, like

```suggestion
    sed -i '/^QTLIB_DIR=/ d' src/Resource_Files/bash/sigil-sh_install
```

ttuegel commented on pkgs/top-level/all-packages.nix:
```diff
@@ -15413,6 +15419,8 @@ in
 
     qca-qt5 = callPackage ../development/libraries/qca-qt5 { };
 
+    qcsxcad = callPackage ../applications/science/electronics/qcsxcad { };
```
Isn't this an application?

freezeboy replied:
Not it is a library, I also had this suprise.
Should I move it to dev/libs ?

doronbehar replied:
It doesn't matter if upstream considers it a library or an application, it matters if we have another qt package that uses it as an input. And we do: `appcsxcad` and `openems`.

freezeboy commented:
I just force-pushed to apply @ttuegel suggestions

freezeboy committed:
amarok: migrate to Qt5.15

freezeboy committed:
communi: migrate to Qt5.15

freezeboy committed:
csound-qt: migrate to Qt5.15

Did not touch the dependency to python27 but may require another PR

freezeboy committed:
fritzing: migrate to Qt5.15

freezeboy committed:
kgraphviewer: migrate to Qt5.15

freezeboy committed:
luminanceHDR: migrate to Qt5.15

freezeboy committed:
rssguard: migrate to Qt5.15

freezeboy committed:
obs-studio,obs-v4l2sink: migrate to Qt5.15

obs-nvi was not touched but should be modified as well,
it requires a custom download that I don't use, so I prefer to
let official maintainer do it.

tobiasBora commented:
Any news on this? Would love to see kmymoney back on my system.

jonringer reviewed (approved):
diff LGTM
I don't want to bitrot much longer.
If there's no regressions from master during my nixpkgs-review, I'll merge

jonringer reviewed (commented):
do you mind rebasing on the lastest master, there's a few conflicts:

```
Auto-merging pkgs/top-level/all-packages.nix
CONFLICT (content): Merge conflict in pkgs/top-level/all-packages.nix
Auto-merging pkgs/top-level/aliases.nix
Auto-merging pkgs/games/pentobi/default.nix
Auto-merging pkgs/development/libraries/qt-5/5.15/default.nix
Removing pkgs/applications/gis/saga/opencv4.patch
Removing pkgs/applications/gis/saga/finite.patch
Removing pkgs/applications/gis/saga/clang_patch.patch
Automatic merge failed; fix conflicts and then commit the result.
https://github.com/NixOS/nixpkgs/pull/102840 failed to build
```

freezeboy commented:
Rebase done

jonringer commented:
LGTM

failures exist on master
```
https://github.com/NixOS/nixpkgs/pull/102840
2 packages marked as broken and skipped:
retroshare retroshare06

3 packages failed to build:
kmymoney obs-ndi xflr5

70 packages built:
amarok appcsxcad birdtray calaos_installer chiaki clipgrab communi csound-qt cutecom digitalbitbox eagle flameshot fritzing golden-cheetah herqq kdiff3 kgraphviewer krename kronometer ktorrent libsForQt5.herqq libsForQt5.kde2-decoration qcsxcad qtcurve libsForQt512.herqq libsForQt512.kde2-decoration libsForQt512.qcsxcad libsForQt512.qtcurve libsForQt514.kde2-decoration libsForQt514.qcsxcad libsForQt514.qtcurve luminanceHDR massif-visualizer minitube obs-move-transition obs-studio obs-v4l2sink obs-wlrobs openems ostinato pentobi peruse plex-media-player protonmail-bridge psi psi-plus python27Packages.python-csxcad python37Packages.python-csxcad python37Packages.python-openems python38Packages.python-csxcad python38Packages.python-openems qlcplus rsibreak rssguard saga samplv1 sdrangel sigil skanlite skrooge solarus solarus-quest-editor sonic-lineup strawberry teamspeak_client tellico virtmanager-qt vnote webcamoid zoom-us
```

freezeboy commented:
I guess `obs-ndi` is normal as the rest of the stack now use qt5.15, I can change its Qt version too, but I won't be able to build it anyway

I start looking the other errors

jonringer commented:
> I guess `obs-ndi` is normal as the rest of the stack now use qt5.15, I can change its Qt version too, but I won't be able to build it anyway
> 
> I start looking the other errors

either way, this is still an improvement. Thanks! :)

doronbehar commented:
> 3 packages failed to build:
> kmymoney obs-ndi xflr5

That's a regression from the last time I tried to build kmymoney - it fails due to xmlsec not building:

https://gist.github.com/doronbehar/4b34941feb73855eb307b3470d12dff2

doronbehar commented:
@freezeboy when you rebased it you forced pushed and forgot some commits of mine :(. For instance qtwebkit was fixed there.

doronbehar commented:
Follow up PR: https://github.com/NixOS/nixpkgs/pull/104474

kmymoney needs https://github.com/NixOS/nixpkgs/pull/104472 in order to work.
